tests/telemetry/test_telemetry_metrics.py

```diff
@@ -0,0 +1,255 @@
+import sys
+
+from mock.mock import ANY
+import pytest
+
+from ddtrace.internal.constants import TELEMETRY_APPSEC
+from ddtrace.internal.constants import TELEMETRY_TRACER
+from ddtrace.internal.constants import TELEMETRY_TYPE_GENERATE_METRICS
+from ddtrace.internal.telemetry.metrics_namespaces import TelemetryTypeError
+from ddtrace.internal.utils.version import _pep440_to_semver
+from tests.telemetry.test_writer import _get_request_body
+
+
+def _assert_metric(test_agent_session_telemetry_metrics, expected_series, namespace=TELEMETRY_TRACER, seq_id=1):
+    test_agent_session_telemetry_metrics.telemetry_writer.periodic()
+    events = test_agent_session_telemetry_metrics.get_events()
+    # TODO: test_agent_session.clear() creates many telemetry events 'app-closing' type
+    assert len([event for event in events if event["request_type"] == TELEMETRY_TYPE_GENERATE_METRICS]) == 1
+
+    payload = {
+        "namespace": namespace,
+        "lib_language": "python",
+        "lib_version": _pep440_to_semver(),
+        "series": expected_series,
+    }
+    assert events[0]["request_type"] == TELEMETRY_TYPE_GENERATE_METRICS
+
+    assert events[0] == _get_request_body(payload, TELEMETRY_TYPE_GENERATE_METRICS, seq_id)
+
+
+@pytest.mark.skipif(sys.version_info < (3, 6), reason="mock.ANY doesn't works in py3.5 or lower")
+def test_send_metric_flush_and_series_is_restarted(test_agent_session_telemetry_metrics, mock_time):
+    """A datapoint is at least: a metric name, a metric value, and the time at which the value was collected.
+    But in Datadog, a datapoint also includes tags, which declare all the various scopes the datapoint belongs to
+    https://www.datadoghq.com/blog/the-power-of-tagged-metrics/#whats-a-metric-tag
+    """
+    telemetry_writer = test_agent_session_telemetry_metrics.telemetry_writer
+    telemetry_writer.add_count_metric(TELEMETRY_TRACER, "test-metric2", 1, {"a": "b"})
+    expected_series = [
+        {
+            "common": True,
+            "host": ANY,
+            "interval": 60,
+            "name": "dd.app_telemetry.tracers.test-metric2",
+            "points": [[1642544540, 1.0]],
+            "tags": {"a": "b"},
+            "type": "count",
+        },
+    ]
+
+    _assert_metric(test_agent_session_telemetry_metrics, expected_series)
+    test_agent_session_telemetry_metrics.clear()
+
+    telemetry_writer.add_count_metric(TELEMETRY_TRACER, "test-metric2", 1, {"a": "b"})
+
+    _assert_metric(test_agent_session_telemetry_metrics, expected_series, seq_id=2)
```

Comment:
I don't understand how this assertion maps onto the idea of "flush and series is restarted". Is the key that the `points` array only has one element in it, indicating that another datapoint was not added to the series?

Reply:
When we add a metric, this data is stored in the telemetry_writer instance. When we run telemetry_writer.periodic() the information is sent to the agent and flush the queue.

We want to test isolated elements in the queue and ensure the queue is empty